Add Pure Storage FlashArray Integration (#1218)
DataDog/integrations-extras: purefa/README.md

# Agent Check: PureFA

## Overview

> This check monitors the [Pure Storage FlashArray][1] through the [Datadog Agent][2] and the [Pure Storage Prometheus exporter][1].

Reviewer: Should "Pure Storage FlashArray" and "Pure Storage Prometheus exporter" point to different links?

Reviewer 2: No - will fix

> This check monitors the [Pure Storage FlashArray][1] through the [Datadog Agent][2] and the [Pure Storage Prometheus exporter][1].

Reviewer: It would also be nice to have a bit more context here about what kind of data this integration displays to the user - i.e. what specific information about their Pure Storage FlashArray will the user have access to in Datadog?

Reviewer 2: Ack

> Note: This integration requires OpenMetricsBaseCheckV2 which is available in Agent v7.26.x+ and requires Python 3

Reviewer: ```suggestion
Note: This integration requires:
-  OpenMetricsBaseCheckV2, which is available in Agent v7.26.x+ 
- Python 3
- Use of the [Pure Storage Prometheus exporter][1]
```

Does the user have to download/do any installation for the Prometheus exporter? If so, that should be mentioned here as well.

Reviewer 2: Ack

> Note: This integration requires the use of the [Pure Storage Prometheus exporter][1]

Reviewer: ```suggestion
```

Reviewer 2: Ack

## Setup

Follow the instructions below to install and configure this check for an Agent running on a host. For containerized environments, see the Autodiscovery Integration Templates for guidance on applying these instructions.

### Installation

1. [Download and launch the Datadog Agent][9].

> 2. Manually install the Pure FlashArray integration. See [Use Community Integrations][10] for more details based on the environment.

Reviewer: ```suggestion
2. Manually install the Pure FlashArray integration. See [Use Community Integrations][10] for more details based on your environment.
```

Reviewer 2: Ack

#### Host

To configure this check for an Agent running on a host, run `datadog-agent integration install -t datadog-purafa==<INTEGRATION_VERSION>`.

### Configuration

> 1. Create a local user on your FlashArray with Read-Only role and generate an API token for this user.

Reviewer: ```suggestion
1. Create a local user on your FlashArray with a Read-Only role and generate an API token for this user.
```

Reviewer 2: Ack

   ![Generate an API Key](https://raw.githubusercontent.com/DataDog/integrations-extras/master/purefa/images/API.png) 

> 2. Add this configuration block to the `purefa.d/conf.yaml` file, in the `conf.d/` folder at the root of your Agent's configuration directory to start collecting your purefa performance data. See the sample [purefa.d/conf.yaml][4] for all available configuration options.

Reviewer: ```suggestion
2. Add the following configuration block to the `purefa.d/conf.yaml` file, in the `conf.d/` folder at the root of your Agent's configuration directory to start collecting your purefa performance data. See the sample [purefa.d/conf.yaml][4] for all available configuration options.
```

Reviewer 2: Ack

## Data Collected

### Metrics

See [metadata.csv][7] for a list of metrics provided by this check.

### Events

The PureFA integration does not include any events.

### Service Checks

The PureFA integration does not include any service checks.

## Support

For support or feature requests, contact Pure Storage through the following methods:
* Email: pure-observability@purestorage.com
* Slack: [Pure Storage Code// Observability Channel][11].

[1]: https://github.com/PureStorage-OpenConnect/pure-exporter
[2]: https://app.datadoghq.com/account/settings#agent
[4]: https://github.com/PureStorage-OpenConnect/observability/blob/master/datadog/integrations-extras/purefa/datadog_checks/purefa/data/conf.yaml.example
[5]: https://docs.datadoghq.com/agent/guide/agent-commands/#start-stop-and-restart-the-agent
[6]: https://docs.datadoghq.com/agent/guide/agent-commands/#agent-status-and-information
[7]: https://github.com/PureStorage-OpenConnect/observability/blob/master/datadog/integrations-extras/purefa/metadata.csv
[9]: https://app.datadoghq.com/account/settings#agent
[10]: https://docs.datadoghq.com/agent/guide/community-integrations-installation-with-docker-agent

> [11]: https://code-purestorage.slack.com/messages/C0357KLR1EU

Reviewer: Are new users able to sign up for your slack? This just brought me to a login page

Reviewer 2: Yes - anyone can sign up for this Slack.

# Agent Check: PureFA

## Overview

This check monitors the [Pure Storage FlashArray][3] through the [Datadog Agent][2] and the [Pure Storage Prometheus exporter][1]. 

The integration can provide performance data at the array, host, volume and pod level, as well as high-level capacity and configuration information.

> Monitor multiple FlashArrays and aggregate these into a single dashboard, or group together by customer defined environment.

Reviewer: ```suggestion
You can now monitor multiple FlashArrays and aggregate these into a single dashboard, or group them together by customer defined environment.
```

Note: This integration requires:

>  - OpenMetricsBaseCheckV2 which is available in Agent v7.26.x+

Reviewer 3: ```suggestion
 - Agent v7.26.x+ to utilize OpenMetricsBaseCheckV2
```

Reviewer 2: Ack

 - Python 3

>  - The [Pure Storage Prometheus exporter][1] is installed and running in a containerized environment. Refer to the [GitHub repo][1] for installation instructions.

Reviewer 3: ```suggestion
 - The Pure Storage Prometheus exporter is installed and running in a containerized environment. Refer to the [GitHub repo][1] for installation instructions.
```
Small nit but both links just go to the same url, so I think removing one of them is more intuitive.

Reviewer 2: Ack

2. Manually install the Pure FlashArray integration. See [Use Community Integrations][10] for more details based on your environment.

#### Host

To configure this check for an Agent running on a host, run `datadog-agent integration install -t datadog-purafa==<INTEGRATION_VERSION>`.

### Configuration

1. Create a local user on your FlashArray with the Read-Only role and generate an API token for this user.
   ![Generate an API Key](https://raw.githubusercontent.com/DataDog/integrations-extras/master/purefa/images/API.png) 
2. Add the following configuration block to the `purefa.d/conf.yaml` file, in the `conf.d/` folder at the root of your Agent's configuration directory to start collecting your purefa performance data. See the sample [purefa.d/conf.yaml][4] for all available configuration options.

Note: The `/array` endpoint is required as an abolute minimum when creating your configuration file.

```yaml
init_config:
   timeout: 60

instances:

  - openmetrics_endpoint: http://<exporter_ip_or_fqdn>:9491/metrics/flasharray/array?endpoint=<array_ip_or_fqdn>
    tags:
       - env:<env>
       - fa_array_name:<full_fqdn>
       - host:<full_fqdn>

>     hostname_label: host
>     hostname_format: <HOSTNAME>.<domain_name>
>     headers:
>        Authorization: Bearer <api_token>
>     min_collection_interval: 120
>     empty_default_hostname: true
>     max_returned_metrics: 100000

Reviewer 3: ```suggestion
```
These are optional and not necessary to include in every set up, so it would be better to remove them from the README.

Reviewer 2: The way our integration works it is better to have these configured this way otherwise strange things can start to happen. for example the api_token is absolutely required for authentication and the max_returned_metrics must be this high to ensure we don't truncate data for very large and heavily configured arrays

Reviewer 2: @yzhan289 are you happy to let us keep this in the README given out requirements?

Reviewer 3: I think the others are fine, although the `max_returned_metrics: 100000` part is something I'll need to check with the team, since it is a hidden config option meant to be used for support cases.

Reviewer 2: That one is actually crucial when we are dealing with enterprise class arrays. The default is way to small

Reviewer 3: Thanks. In that case, can you have the config enabled and uncommented in the `conf.yaml`?

Reviewer 2: @yzhan289 can you please re-review this PR to clear your change request?

  - openmetrics_endpoint: http://<exporter_ip_or_fqdn>:9491/metrics/flasharray/volumes?endpoint=<array_ip_or_fqdn>
    tags:
       - env:<env>
       - fa_array_name:<full_fqdn>

>     hostname_label: host
>     hostname_format: <HOSTNAME>.<domain_name>
>     headers:
>        Authorization: Bearer <api_token>
>     min_collection_interval: 120
>     empty_default_hostname: true
>     max_returned_metrics: 100000

Reviewer 3: ```suggestion
```

Reviewer 2: As above

  - openmetrics_endpoint: http://<exporter_ip_or_fqdn>:9491/metrics/flasharray/hosts?endpoint=<array_ip_or_fqdn>
    tags:
       - env:<env>
       - fa_array_name:<full_fqdn>

>     hostname_label: host
>     hostname_format: <HOSTNAME>.<domain_name>
>     headers:
>        Authorization: Bearer <api_token>
>     min_collection_interval: 120
>     empty_default_hostname: true
>     max_returned_metrics: 100000

Reviewer 3: ```suggestion
```

Reviewer 2: As above

  - openmetrics_endpoint: http://<exporter_ip_or_fqdn>:9491/metrics/flasharray/pods?endpoint=<array_ip_or_fqdn>
    tags:
       - env:<env>
       - fa_array_name:<full_fqdn>
       - host:<full_fqdn>

>     hostname_label: host
>     hostname_format: <HOSTNAME>.<domain_name>
>     headers:
>        Authorization: Bearer <api_token>
>     min_collection_interval: 120
>     empty_default_hostname: true
>     max_returned_metrics: 100000

Reviewer 3: ```suggestion
```

Reviewer 2: As above

```
2. [Restart the Agent][5].

### Validation

[Run the Agent's status subcommand][6] and look for `purefa` under the Checks section.

## Data Collected

### Metrics

See [metadata.csv][7] for a list of metrics provided by this check.

### Events

The PureFA integration does not include any events.

### Service Checks

The PureFA integration does not include any service checks.

## Support

For support or feature requests, contact Pure Storage through the following methods:
* Email: pure-observability@purestorage.com

> * Slack: [Pure Storage Code// Observability Channel][11].

Reviewer: I believe this section is not necessary - for integrations on the main Integrations page, Datadog is listed as the support contact automatically.

Reviewer 2: @anawishnoff question here for you...

Reviewer: @sdodsley Apologies for the confusion here, I see now that you're using manifest v2 so the structure of the tile is different (a Support tab with your custom content will be included on your tile)- you can safely ignore my initial comment and keep your support info here. Thanks!

# Agent Check: PureFA

## Overview

This check monitors the [Pure Storage FlashArray][3] through the [Datadog Agent][2] and the [Pure Storage Prometheus exporter][1]. 

The integration can provide performance data at the array, host, volume and pod level, as well as high-level capacity and configuration information.

You can monitor multiple FlashArrays and aggregate these into a single dashboard, or group them together by customer defined environment.

> Note: This integration requires:

Reviewer 4: ```suggestion
**This integration requires the following**:
```

 - Agent v7.26.x+ to utilize OpenMetricsBaseCheckV2
 - Python 3
 - The Pure Storage Prometheus exporter is installed and running in a containerized environment. Refer to the [GitHub repo][1] for installation instructions.

## Setup

Follow the instructions below to install and configure this check for an Agent running on a host. For containerized environments, see the Autodiscovery Integration Templates for guidance on applying these instructions.

### Installation

1. [Download and launch the Datadog Agent][9].
2. Manually install the Pure FlashArray integration. See [Use Community Integrations][10] for more details based on your environment.

#### Host

To configure this check for an Agent running on a host, run `datadog-agent integration install -t datadog-purefa==<INTEGRATION_VERSION>`.

### Configuration

1. Create a local user on your FlashArray with the Read-Only role and generate an API token for this user.
   ![Generate an API Key](https://raw.githubusercontent.com/DataDog/integrations-extras/master/purefa/images/API.png) 

> 2. Add the following configuration block to the `purefa.d/conf.yaml` file, in the `conf.d/` folder at the root of your Agent's configuration directory to start collecting your purefa performance data. See the sample [purefa.d/conf.yaml][4] for all available configuration options.

Reviewer 4: ```suggestion
2. Add the following configuration block to the `purefa.d/conf.yaml` file, in the `conf.d/` folder at the root of your Agent's configuration directory to start collecting your PureFA performance data. See the sample [purefa.d/conf.yaml][4] for all available configuration options.
```

> Note: The `/array` endpoint is required as an absolute minimum when creating your configuration file.

Reviewer 4: ```suggestion
**Note**: The `/array` endpoint is required as an absolute minimum when creating your configuration file.
```

       - env:<env>
       - fa_array_name:<full_fqdn>
       - host:<full_fqdn>
    headers:
       Authorization: Bearer <api_token>
    min_collection_interval: 120

  - openmetrics_endpoint: http://<exporter_ip_or_fqdn>:<port>/metrics/flasharray/volumes?endpoint=<array_ip_or_fqdn>
    tags:
       - env:<env>
       - fa_array_name:<full_fqdn>
    headers:
       Authorization: Bearer <api_token>
    min_collection_interval: 120

  - openmetrics_endpoint: http://<exporter_ip_or_fqdn>:<port>/metrics/flasharray/hosts?endpoint=<array_ip_or_fqdn>
    tags:
       - env:<env>
       - fa_array_name:<full_fqdn>
    headers:
       Authorization: Bearer <api_token>
    min_collection_interval: 120

  - openmetrics_endpoint: http://<exporter_ip_or_fqdn>:<port>/metrics/flasharray/pods?endpoint=<array_ip_or_fqdn>
    tags:
       - env:<env>
       - fa_array_name:<full_fqdn>
       - host:<full_fqdn>
    headers:
       Authorization: Bearer <api_token>
    min_collection_interval: 120
```

2. [Restart the Agent][5].

### Validation

[Run the Agent's status subcommand][6] and look for `purefa` under the Checks section.

### Troubleshooting

#### Arrays are not showing in dashboard

> The dashboards included in this integration use the tags `env`, and `fa_array_name`. Please make sure that these are set per instance. `host` must also be set for the `/array` and `/pods` endpoints in `purefa.d/conf.yaml`

Reviewer 4: ```suggestion
The dashboards included in this integration use the tags `env`, and `fa_array_name`. Make sure that these are set per instance. `host` must also be set for the `/array` and `/pods` endpoints in `purefa.d/conf.yaml`.
```

```yaml
- tags:
   - env:<env>
   - fa_array_name:<full_fqdn>
   - host:<full_fqdn>
```

> #### Increasing Collection Interval

Reviewer 4: ```suggestion
#### Increasing collection interval
```

    tags:
       - env:<env>
       - fa_array_name:<full_fqdn>
    headers:
       Authorization: Bearer <api_token>

  - openmetrics_endpoint: http://<exporter_ip_or_fqdn>:9491/metrics/flasharray/pods?endpoint=<array_ip_or_fqdn>
    tags:
       - env:<env>
       - fa_array_name:<full_fqdn>
       - host:<full_fqdn>
    headers:
       Authorization: Bearer <api_token>
```
2. [Restart the Agent][5].

### Validation

[Run the Agent's status subcommand][6] and look for `purefa` under the Checks section.

## Data Collected

### Metrics

See [metadata.csv][7] for a list of metrics provided by this check.

### Events

The PureFA integration does not include any events.

### Service Checks

The PureFA integration does not include any service checks.

## Troubleshooting

### Arrays are not showing in dashboard

The dashboards included in this integration use the tags `env`, and `fa_array_name`. Please make sure that these are set per instance. `host` must also be set for the `/array` and `/pods` endpoints in `purefa.d/conf.yaml`

```yaml
- tags:
   - env:<env>
   - fa_array_name:<full_fqdn>
   - host:<full_fqdn>
```

### Increasing Collection Interval

> The Pure Storage FlashArray check sets `min_collection_interval` to `120` by default, the minimum recomended value is `20`. You may increase/decrease `min_collection_interval` in `purefa.d/conf.yaml` file if necessary:

Reviewer: ```suggestion
The Pure Storage FlashArray check sets `min_collection_interval` to `120` by default, and the minimum recommended value is `20`. You may increase/decrease `min_collection_interval` in the `purefa.d/conf.yaml` file if necessary:
```

    headers:
       Authorization: Bearer <api_token>
    min_collection_interval: 120

  - openmetrics_endpoint: http://<exporter_ip_or_fqdn>:<port>/metrics/flasharray/pods?endpoint=<array_ip_or_fqdn>
    tags:
       - env:<env>
       - fa_array_name:<full_fqdn>
       - host:<full_fqdn>
    headers:
       Authorization: Bearer <api_token>
    min_collection_interval: 120
```

2. [Restart the Agent][5].

### Validation

[Run the Agent's status subcommand][6] and look for `purefa` under the Checks section.

### Troubleshooting

#### Arrays are not showing in dashboard

The dashboards included in this integration use the tags `env`, and `fa_array_name`. Please make sure that these are set per instance. `host` must also be set for the `/array` and `/pods` endpoints in `purefa.d/conf.yaml`

```yaml
- tags:
   - env:<env>
   - fa_array_name:<full_fqdn>
   - host:<full_fqdn>
```

#### Increasing Collection Interval

The Pure Storage FlashArray check sets `min_collection_interval` to `120` by default, and the minimum recommended value is `20`. You may increase/decrease `min_collection_interval` in the `purefa.d/conf.yaml` file if necessary:

```yaml
min_collection_interval: 120
```

> #### Missing metrics
> 
> Since there are many important metrics in a Pure Storage Array, the Pure Storage FlashArray check sets `max_returned_metrics` to `100000` by default. You may increase/decrease `max_returned_metrics` in the `purefa.d/conf.yaml` file if necessary:
> 
> ```yaml
> max_returned_metrics: 100000
> ```

Reviewer 5: ```suggestion
```
Let's delete this since it's the default in the code now.
